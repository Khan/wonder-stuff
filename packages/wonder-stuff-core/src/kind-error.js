// @flow
import {Errors} from "./errors.js";
import {ErrorInfo} from "./error-info.js";
import {cloneMetadata} from "./clone-metadata.js";

import type {Metadata} from "./types.js";

/**
 * Options for constructing a `KindError`.
 */
type Options = {|
    /**
     * An error responsible for the error being created.
     *
     * @type {?Error}
     */
    cause?: ?Error,

    /**
     * Data to be attached to the error.
     *
     * @type {?$ReadOnly<Metadata>}
     */
    metadata?: ?$ReadOnly<Metadata>,

    /**
     * A prefix for the error name.
     *
     * @type {?string}
     */
    prefix?: ?string,

    /**
     * A name for the error.
     *
     * @type {?string}
     */
    name?: ?string,

    /**
     * The number of stack frames to strip from the error.
     *
     * @type {?number}
     */
    stripStackFrames?: ?number,

    /**
     * The minimum number of stack frames to try and retain.
     *
     * @type {?number}
     */
    minimumFrameCount?: ?number,
|};

/**
 * An error to describe detailed states and relationships.
 *
 * This error type supports error taxonomies and attachment of metadata, as
 * well as causal relationships between itself and another source error.
 * More specific variations can be built off this to provide a payload for
 * logging setups like Sentry or winston.
 */
export class KindError extends Error {
    +kind: string;
    +originalMessage: string;
    +metadata: ?$ReadOnly<Metadata>;
    +cause: ?Error;

    /**
     * Creates an instance of `KindError`.
     *
     * @memberof KindError
     * @param {string} message The error message.
     * @param {string} [kind] The kind of error. This will be combined with
     * `prefix` to form the name of the error, i.e. PrefixKindError.
     * Defaults to `Errors.Unknown`.
     * @param {Options} [options] Options for constructing the error.
     * @param {Error} [options.cause] The error that caused this error.
     * @param {$ReadOnly<Metadata>} [options.metadata] The metadata to attach
     * to the error.
     * @param {string} [options.prefix=""] A prefix to prepend the name of the
     * error.
     * @param {string} [options.name="Error"] The name of the error.
     * @param {number} [options.stripStackFrames=0] The number of stack frames
     * to remove from the error's stack. This can be used to ensure that the top
     * call of the stack references the point at which an error is thrown which
     * can be useful when helper functions are used to build the error being
     * thrown.
     * @param {number} [options.minimumFrameCount=1] The minimum number of
     * stack frames to try and retain. This can be used to prevent stripping
     * all stack frames from the error's stack.
     */
    constructor(
        message: string,
        kind: string = Errors.Unknown,
        {
            cause,
            prefix,
            name,
            metadata,
            stripStackFrames,
            minimumFrameCount,
        }: Options = {},
    ) {
        // Validate arguments.
        if (cause && !(cause instanceof Error)) {
            throw new Error("cause must be an instance of Error");
        }
        if (name != null && /\s/g.test(name)) {
            throw new Error("name must not contain whitespace");
        }
        if (/\s/g.test(kind)) {
            throw new Error("kind must not contain whitespace");
        }
        if (prefix != null && /\s/g.test(prefix)) {
            throw new Error("prefix must not contain whitespace");
        }
        if (stripStackFrames != null && stripStackFrames < 0) {
            throw new Error("stripStackFrames must be >= 0");
        }
        if (minimumFrameCount != null && minimumFrameCount < 0) {
            throw new Error("minimumFrameCount must be >= 0");
        }

        super(message);

        // Save the original message as we may change this message later.
        this.originalMessage = message;

        // Metadata associated with this error.
        this.metadata = cloneMetadata(metadata);

        // Set the name so we get a nice error output, like
        // KAInternalError
        this.name = `${prefix ?? ""}${kind}${name ?? ""}Error`;

        // The kind of error which can be used for grouping with
        // other error sources that use the same error taxonomy.
        this.kind = kind;

        // The cause of this error, if there is one.
        this.cause = cause;
        if (cause != null) {
            // We want to generate a better stack trace using the source error
            // and our own stack.
            const normalizedError = ErrorInfo.normalize(
                this,
                stripStackFrames ?? 0,
                minimumFrameCount ?? 1,
            );
            // We don't normalize the cause as this should have happened already
            // when it was, itself, created with a cause.
            const normalizedCause = ErrorInfo.from(cause);
            const combined = ErrorInfo.fromConsequenceAndCause(
                normalizedError,
                normalizedCause,
            );

            /**
             * $FlowIgnore[incompatible-type]
             * Flow doesn't like us leaving KAError temporarily without a
             * stack, but I don't want this crashing because we try to
             * assign to a readonly thing, so being mildly cautious.
             */
            delete this.stack;
            this.stack = combined.toString();

            // We update our message to a normalized one from the combined
            // stacks, as this gives us a nice causal relationship.
            this.message = combined.message;
        }
    }
}
