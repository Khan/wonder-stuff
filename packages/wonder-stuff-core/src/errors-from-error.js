// @flow
import {KindError} from "./kind-error.js";
import {Errors} from "./errors.js";

export const Order = Object.freeze({
    ConsequenceFirst: "consequence-first",
    CauseFirst: "cause-first",
});

export type SequenceOrder = $Values<typeof Order>;

/**
 * Given an error, generates a sequence of `Error`s.
 *
 * If the given error is a `KindError`-derivitive, this function will return
 * a sequence starting with that error and then including any causal errors,
 * chaining through causal `KindError`s accordingly.
 *
 * If `reverse` is `true`, the generated sequence starts with the lowermost
 * causal error and ends with the given error.
 */
export function* errorsFromError(
    error: ?Error,
    order: SequenceOrder,
): Iterator<Error> {
    if (order !== Order.CauseFirst && order !== Order.ConsequenceFirst) {
        throw new KindError("Invalid sequence order", Errors.InvalidInput, {
            metadata: {
                order,
            },
        });
    }
    if (error == null) {
        return;
    }

    if (order === Order.ConsequenceFirst) {
        yield error;
    }
    if (error instanceof KindError) {
        yield* errorsFromError(error.cause, order);
    }
    if (order === Order.CauseFirst) {
        yield error;
    }
}
