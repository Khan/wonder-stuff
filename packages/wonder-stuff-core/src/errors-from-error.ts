import {KindError} from "./kind-error";
import {Errors} from "./errors";

export const Order = Object.freeze({
    ConsequenceFirst: "consequence-first",
    CauseFirst: "cause-first",
});

export type SequenceOrder = typeof Order[keyof typeof Order];

/**
 * Given an error, generates a sequence of `Error`s.
 *
 * If the given error is a `KindError`-derivitive, this function will return
 * a sequence starting with that error and then including any causal errors,
 * chaining through causal `KindError`s accordingly.
 *
 * If `reverse` is `true`, the generated sequence starts with the lowermost
 * causal error and ends with the given error.
 */
export function* errorsFromError(error: Error | null | undefined, order: SequenceOrder): Iterator<Error> {
    if (order !== Order.CauseFirst && order !== Order.ConsequenceFirst) {
        throw new KindError("Invalid sequence order", Errors.InvalidInput, {
            metadata: {
                order,
            },
        });
    }
    if (error == null) {
        return;
    }

    if (order === Order.ConsequenceFirst) {
        yield error;
    }
    if (error instanceof KindError) {
// @ts-expect-error - TS2488 - Type 'Iterator<Error, any, undefined>' must have a '[Symbol.iterator]()' method that returns an iterator.
        yield* errorsFromError(error.cause, order);
    }
    if (order === Order.CauseFirst) {
        yield error;
    }
}
