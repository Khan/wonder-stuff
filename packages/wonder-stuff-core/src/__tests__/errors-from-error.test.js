// @flow
import {KindError} from "../kind-error.js";
import {errorsFromError, Order} from "../errors-from-error.js";

describe("#errorsFromError", () => {
    it.each([null, undefined])(
        "should yield an empty sequence if the error is nullish",
        (error) => {
            // Arrange

            // Act
            const result = Array.from(
                errorsFromError(error, Order.ConsequenceFirst),
            );

            // Assert
            expect(result).toBeEmpty();
        },
    );

    it("should yield just the error in sequence if it is not a KindError derivative", () => {
        // Arrange
        const error = new Error("test");

        // Act
        const result = Array.from(
            errorsFromError(error, Order.ConsequenceFirst),
        );

        // Assert
        expect(result).toEqual([error]);
    });

    describe("when Order.ConsequenceFirst", () => {
        it("should yield errors in the order of the root error down to the lowermost cause", () => {
            // Arrange
            class MyError extends KindError {
                constructor(message: string, options: {cause?: Error}) {
                    super(message, "CustomKind", {...options});
                }
            }
            const cause3 = new Error("cause2");
            const cause2 = new MyError("cause2", {cause: cause3});
            const cause1 = new KindError("cause1", "Kind1", {cause: cause2});
            const rootError = new KindError("root", "RootKind", {
                cause: cause1,
            });

            // Act
            const result = Array.from(
                errorsFromError(rootError, Order.ConsequenceFirst),
            );

            // Assert
            expect(result).toEqual([rootError, cause1, cause2, cause3]);
        });
    });

    describe("when Order.CauseFirst", () => {
        it("should yield errors in the order of the lowermost causal error up to the root consequence", () => {
            // Arrange
            class MyError extends KindError {
                constructor(message: string, options: {cause?: Error}) {
                    super(message, "CustomKind", {...options});
                }
            }
            const cause3 = new Error("cause2");
            const cause2 = new MyError("cause2", {cause: cause3});
            const cause1 = new KindError("cause1", "Kind1", {cause: cause2});
            const rootError = new KindError("root", "RootKind", {
                cause: cause1,
            });

            // Act
            const result = Array.from(
                errorsFromError(rootError, Order.CauseFirst),
            );

            // Assert
            expect(result).toEqual([cause3, cause2, cause1, rootError]);
        });
    });
});
