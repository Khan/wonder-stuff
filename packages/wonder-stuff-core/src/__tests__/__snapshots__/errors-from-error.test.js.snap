// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#errorsFromError should throw if the order is invalid 1`] = `"Invalid sequence order"`;

exports[`#errorsFromError should throw if the order is invalid 2`] = `"Invalid sequence order"`;

exports[`#errorsFromError should throw if the order is invalid 3`] = `"Invalid sequence order"`;

exports[`#errorsFromError should throw if the order is invalid 4`] = `"Invalid sequence order"`;
